### PR TITLE
Neighbor search using tree

### PR DESCRIPTION
**Pull request by maxscheurer**
Use `NearestNeighbor.jl` for faster range-based neighbor search.

## ToDos
- [x] real unit tests
- [x] how to set leaf size (sensible default)?

I've tested the code using the following script:
```Julia
using Molly
using BenchmarkTools
using Test

temp = 298
timestep = 0.002
n_steps = 20_000
box_size = 25.0

n_atoms = 40000
coords = [box_size .* rand(SVector{3}) for i in 1:(n_atoms / 2)]
for i in 1:length(coords)
    push!(coords, coords[i] .+ [0.1, 0.0, 0.0])
end
bonds = [HarmonicBond(i, i + (n_atoms ÷ 2), 0.1, 300_000.0) for i in 1:(n_atoms ÷ 2)]

s = Simulation(
    simulator=VelocityVerlet(),
    atoms=[Atom(attype="H", name="H", resnum=i, resname="H", charge=0.0,
                mass=10.0, σ=0.3, ϵ=0.2) for i in 1:n_atoms],
    specific_inter_lists=(bonds,),
    general_inters=(LennardJones(true),),
    coords=coords,
    velocities=[velocity(10.0, temp) .* 0.01 for i in 1:n_atoms],
    temperature=temp,
    box_size=box_size,
    neighbour_finder=DistanceNeighbourFinder(trues(n_atoms, n_atoms), 10, 2.0),
    thermostat=AndersenThermostat(10.0),
    loggers=Dict("temp" => TemperatureLogger(10),
                    "coords" => CoordinateLogger(10)),
    timestep=timestep,
    n_steps=n_steps
)

nf_tree = DistanceNeighbourFinderTree(trues(n_atoms, n_atoms), 10, 2.0)

# Test
find_neighbours!(s, s.neighbour_finder, 0; parallel = true)
ref = copy(s.neighbours)
find_neighbours!(s, nf_tree, 0; parallel = true)
@test s.neighbours == ref

# Benchmark
tree_parallel = @benchmark find_neighbours!(s, nf_tree, 0; parallel = true)
tree_serial = @benchmark find_neighbours!(s, nf_tree, 0; parallel = false)

dist_parallel = @benchmark find_neighbours!(s, s.neighbour_finder, 0; parallel = true)
dist_serial = @benchmark find_neighbours!(s, s.neighbour_finder, 0; parallel = false)
```

which yields the following benchmarks on my machine (2 threads for parallel runs):
```Julia
julia> @show dist_serial
dist_serial = Trial(14.153 s)
BenchmarkTools.Trial:
  memory estimate:  198.97 MiB
  allocs estimate:  80001
  --------------
  minimum time:     14.153 s (0.00% GC)
  median time:      14.153 s (0.00% GC)
  mean time:        14.153 s (0.00% GC)
  maximum time:     14.153 s (0.00% GC)
  --------------
  samples:          1
  evals/sample:     1

julia> @show tree_serial
tree_serial = Trial(2.320 s)
BenchmarkTools.Trial:
  memory estimate:  285.76 MiB
  allocs estimate:  358942
  --------------
  minimum time:     2.320 s (4.27% GC)
  median time:      2.505 s (1.98% GC)
  mean time:        2.505 s (1.98% GC)
  maximum time:     2.690 s (0.00% GC)
  --------------
  samples:          2
  evals/sample:     1

julia> @show dist_parallel
dist_parallel = Trial(11.257 s)
BenchmarkTools.Trial:
  memory estimate:  240.98 MiB
  allocs estimate:  80054
  --------------
  minimum time:     11.257 s (0.00% GC)
  median time:      11.257 s (0.00% GC)
  mean time:        11.257 s (0.00% GC)
  maximum time:     11.257 s (0.00% GC)
  --------------
  samples:          1
  evals/sample:     1

julia> @show tree_parallel
tree_parallel = Trial(1.065 s)
BenchmarkTools.Trial:
  memory estimate:  327.76 MiB
  allocs estimate:  358995
  --------------
  minimum time:     1.065 s (0.00% GC)
  median time:      1.087 s (0.00% GC)
  mean time:        1.136 s (3.35% GC)
  maximum time:     1.303 s (14.62% GC)
  --------------
  samples:          5
  evals/sample:     1
```

I'm sure there are still things to optimize and fine-tune here, but the speed-ups actually look quite good.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -7,9 +7,11 @@ version = "0.2.1"
 BioStructures = "de9282ab-8554-53be-b2d6-f6c222edabfc"
 CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
 Colors = "5ae59095-9a9b-59fe-a467-6f913c188581"
+Distances = "b4f34e82-e78d-54a5-968a-f98e89d6e8f7"
```
Comment (jgreener64) [inline Project.toml:10]: Add compat bounds to this file too, e.g. 0.4 for NearestNeighbors and 0.10 for Distances.

File: src/neighbours.jl [hunk 2]
```diff
@@ -85,3 +86,65 @@ function find_neighbours!(s::Simulation,
         end
     end
 end
+
+
+"""
+    DistanceNeighbourFinderTree(nb_matrix, n_steps, dist_cutoff)
```
Comment (jgreener64) [inline src/neighbours.jl:92]: Rename to `DistanceTreeNeighbourFinder` or similar? I kind of like the convention of finishing all neighbour finders with `NeighbourFinder`.

File: src/neighbours.jl [hunk 3]
```diff
@@ -85,3 +86,65 @@ function find_neighbours!(s::Simulation,
...
+            nbi = nf.nb_matrix[:, i]
+            idxs = inrange(btree, ci, nf.dist_cutoff, true)
+            for j in idxs
+                if nbi[j] && i > j
+                    push!(neighbours, (i, j))
+                end
+            end
+        end
+    end
+end
```
Comment (jgreener64) [inline src/neighbours.jl:150]: Add newline.

**Review by jgreener64 — COMMENTED** [review]
This looks great, thanks. I can reproduce a similar speedup on my machine. I have made a few suggestions but this is almost ready to go in.

For the leaf size, I don't have much of a feel for how important it is. If you think it could be an argument to the neighbour finder with a sensible default then feel free to do that, otherwise just leave it to the default settings.

Regarding tests, you could add something similar to your 4 testing lines around line 107 of `test/runtests.jl`.

Also a reference to the new type in line 586 of `docs/src/docs.md` would be good.

File: src/neighbours.jl (hunk 2 above)
Comment (maxscheurer) [reply src/neighbours.jl:92]: Or just `TreeNeighbourFinder` maybe?

Comment (jgreener64) [reply src/neighbours.jl:92]: Sounds good.

File: src/neighbours.jl (hunk 3 above)
Comment (maxscheurer) [reply src/neighbours.jl]: Done.

File: Project.toml (hunk 1 above)
Comment (maxscheurer) [reply Project.toml:10]: Done.

**Comment by maxscheurer** [comment]
I think for now the default value should be alright. I'll probably try to run some benchmarks on different leaf sizes when I have time. 😄

**Comment by jgreener64** [comment]
Merged, thanks a lot.

**Comment by maxscheurer** [comment]
Yay, my first contribution to a Julia project! Hopefully not the last.. 😉 😄

**Comment by jgreener64** [comment]
Absolutely, keep contributing to this project if you have the time!

**Pull request merged**